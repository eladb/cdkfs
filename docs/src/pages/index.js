import React from 'react';
import clsx from 'clsx';
import Link from '@docusaurus/Link';
import useDocusaurusContext from '@docusaurus/useDocusaurusContext';
import Layout from '@theme/Layout';

import styles from './index.module.css';

function HomepageHeader() {
  const {siteConfig} = useDocusaurusContext();
  return (
    <header className={clsx('hero hero--primary', styles.heroBanner)}>
      <div className="container">
        <h1 className="hero__title">{siteConfig.title}</h1>
        <p className="hero__subtitle">{siteConfig.tagline}</p>
        <iframe width="560" height="315" src="https://www.youtube.com/embed/SOWMPzXtTCw" title="YouTube video player"
                frameBorder="0"
                allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share"
                allowFullScreen></iframe>
      </div>
    </header>
  );
}

export default function Home() {
  const {siteConfig} = useDocusaurusContext();
  return (
    <Layout
      title={`Hello from ${siteConfig.title}`}
      >
      <HomepageHeader />
      <main>
      </main>
    </Layout>
  );
}
