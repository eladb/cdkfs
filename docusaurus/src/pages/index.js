import React from 'react';
import useDocusaurusContext from '@docusaurus/useDocusaurusContext';
import Layout from '@theme/Layout';
import Image1 from '../../static/img/backgrounds/1.png';
import Image2 from '../../static/img/backgrounds/2.png';
import TypeScriptLogo from '../../static/img/languages/ts.png';
import GoLogo from '../../static/img/languages/go.png';
import JavaLogo from '../../static/img/languages/java.png';
import PythonLogo from '../../static/img/languages/python.png';

export default function Home() {
    const {siteConfig} = useDocusaurusContext();
    return (
        <Layout
            title={siteConfig.title}
        >
            <main>
                {/* SECTION 1 */}
                <section className="bg-slate-800 text-white py-16 lg:py-36 relative overflow-hidden">
                    <div className="container mx-auto text-center">
                        <img src={siteConfig.themeConfig.image}
                             alt="The projen logo containing three boxes stacked in a pyramid shape with the bottom right box having the letters 'pj' in it"
                             className="mx-auto mb-8 lg:w-36 relative z-10"/>
                        <h1 className="text-6xl font-bold mb-8 relative z-10">{siteConfig.title}</h1>
                        <div className="container mx-auto text-center relative z-10">
                            <h2 className="text-2xl lg:text-5xl font-medium text-white mb-16 lg:mb-24">
                                Rapidly build modern applications
                                <span className="lg:block mt-3"> with advanced configuration management</span>
                            </h2>
                            <a href={`${siteConfig.baseUrl}docs/getting-started/`}
                               className="rounded text-white bg-blue-500 hover:text-white py-3 px-8 inline-block lg:mb-36">Get
                                Started</a>
                        </div>
                        <img src={Image1}
                             alt="An illustration containing various lines, shapes in blues, reds and whites making a larger cube centered inside of a circular pattern."
                             className="mx-auto mb-8 opacity-25 absolute -left-48 top-24 lg:top-48 z-0"/>

                    </div>
                </section>

                <hr className="m-0 bg-slate-600"/>

                {/* SECTION 2 */}
                <section className="dark:bg-slate-800 pt-16 pb-36 lg:py-36">
                    <div className="p-8 lg:p-0 lg:container lg:mx-auto text-center relative z-10">
                        <h2 className="text-2xl lg:text-5xl font-medium dark:text-white mb-16 lg:mb-24">
                            Build and manage your
                            <span className="lg:block mt-3"> apps with confidence</span>
                        </h2>

                        <div className="lg:grid lg:grid-cols-4">
                            <div className="lg:col-start-2 lg:col-span-2">
                                <p className="text-xl lg:text-2xl text-center">
                                    Synthesize project configuration files from a well-typed definitions, written in
                                    {' '}<a href="https://github.com/aws/jsii" target="_blank"
                                            className="text-blue-500 dark:text-blue-300 underline">jsii</a>{' '}
                                    compatible languages reducing project configuration management fatigue, increasing
                                    productivity and confidence.
                                </p>
                            </div>
                        </div>
                    </div>
                </section>

                <hr className="m-0 bg-slate-600"/>

                {/* SECTION 3 */}
                <section className="dark:bg-slate-800 pt-16 pb-36 lg:py-36">
                    <div className="p-8 lg:p-0 lg:container lg:mx-auto text-center relative z-10">
                        <h2 className="text-2xl lg:text-5xl font-medium dark:text-white mb-16 lg:mb-24">
                            Available for these
                            <span className="lg:block mt-3"> popular languages</span>
                        </h2>

                        <ul className="lg:grid lg:grid-cols-5 lg:gap-4 lg:items-center p-0">
                            <li className="list-none mb-12 lg:mb-0 text-center">
                                <img src={TypeScriptLogo} alt="TypeScript" className="h-48"/>
                            </li>
                            <li className="list-none mb-12 lg:mb-0 text-center">
                                <img src={GoLogo} alt="Go" className="h-48 w-auto"/>
                            </li>
                            <li className="list-none mb-12 lg:mb-0 text-center">
                                <img src={JavaLogo} alt="Java" className="h-48 w-auto"/>
                            </li>
                            <li className="list-none mb-12 lg:mb-0 text-center">
                                <img src={PythonLogo} alt="Python" className="rounded"/>
                            </li>
                        </ul>
                    </div>
                </section>

                <hr className="m-0 bg-slate-600"/>

                {/* SECTION 4 */}
                <section className="dark:bg-slate-800 pt-16 pb-36 lg:py-36 relative overflow-hidden">
                    <div className="p-8 lg:p-0 lg:container lg:mx-auto text-center relative z-10">
                        <h2 className="text-2xl lg:text-5xl font-medium dark:text-white mb-16 lg:mb-24">
                            Extensible, customizable
                            <span className="lg:block mt-3"> and adaptable</span>
                        </h2>

                        <div className="lg:grid lg:grid-cols-4">
                            <div className="lg:col-start-2 lg:col-span-2">
                                <p className="text-xl lg:text-2xl text-center">
                                    Tasks allow you to define commands backed by shell scripts. You can use tasks to implement custom
                                    workflows accessible through the projen CLI. Focus more on business requirements,
                                    less on managing project configuration.
                                </p>
                            </div>
                        </div>
                    </div>
                    <img src={Image2}
                         alt="A vector illustration of a cube made from red, yellow and blue puzzle pieces."
                         className="mx-auto mb-8 opacity-[.10] lg:opacity-25 absolute -right-48 top-24 lg:top-48 z-0"/>
                </section>

                <hr className="m-0 bg-slate-600"/>

                {/* SECTION 5 */}
                <section className="dark:bg-slate-800 pt-16 pb-36 lg:py-36 relative overflow-hidden">
                    <div className="p-8 lg:p-0 lg:container lg:mx-auto text-center relative z-10">
                        <h2 className="text-2xl lg:text-5xl font-medium dark:text-white mb-16 lg:mb-24">
                            Built by developers,
                            <span className="lg:block mt-3"> for developers</span>
                        </h2>

                        <div className="lg:grid lg:grid-cols-4 mb-8">
                            <div className="lg:col-start-2 lg:col-span-2">
                                <p className="text-xl lg:text-2xl text-center">
                                    Youâ€™re not alone. Developers from all over the world manage their complex configurations using
                                    projen. <a href={`${siteConfig.baseUrl}docs/getting-started/`} className="text-blue-500 dark:text-blue-300 underline">Get started</a> today
                                    and get these benefits:
                                </p>
                            </div>
                        </div>

                        <ul className="md:grid md:grid-cols-12 md:gap-4 pl-0">
                            <li className="list-none md:col-span-6 rounded border-solid dark:border-slate-600 dark:bg-slate-700 p-6 mb-8 lg:mb-0">
                                Standard npm scripts like compile, build, test, package and more.
                            </li>
                            <li className="list-none md:col-span-6 rounded border-solid dark:border-slate-600 dark:bg-slate-700 p-6 mb-8 lg:mb-0">
                                jsii: compile, package, api compatibility checks, API.md
                            </li>
                            <li className="list-none md:col-span-4 rounded border-solid dark:border-slate-600 dark:bg-slate-700 p-6 mb-8 lg:mb-0">
                                Fully synthesize package.json
                            </li>
                            <li className="list-none md:col-span-4 rounded border-solid dark:border-slate-600 dark:bg-slate-700 p-6 mb-8 lg:mb-0">
                                Automated PR builds
                            </li>
                            <li className="list-none md:col-span-4 rounded border-solid dark:border-slate-600 dark:bg-slate-700 p-6 mb-8 lg:mb-0">
                                Automated dependency upgrades
                            </li>
                            <li className="list-none col-span-6 rounded border-solid dark:border-slate-600 dark:bg-slate-700 p-6 mb-8 lg:mb-0">
                                Bump & release scripts with CHANGELOG generation based on conventional commits
                            </li>
                            <li className="list-none col-span-6 rounded border-solid dark:border-slate-600 dark:bg-slate-700 p-6 mb-8 lg:mb-0">
                                Node "engines" support with coupling to CI build environment and @types/node
                            </li>
                        </ul>
                    </div>
                </section>
            </main>
        </Layout>
    );
}
