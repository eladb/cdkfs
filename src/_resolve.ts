
export function resolve(value: any, ...args: any[]): any {
  if (value == null) {
    return value;
  }

  // if value is a function, call it and resolve the result.
  if (typeof(value) === 'function') {
    const resolved = value.apply(undefined, args);
    return resolve(resolved, ...args);
  }

  if (typeof(value) !== 'object') {
    return value;
  }

  if (Array.isArray(value)) {
    return value.map(x => resolve(x, ...args));
  }

  const result: any = {};

  for (const [k,v] of Object.entries(value)) {
    result[k] = resolve(v, ...args);
  }

  return result;
}