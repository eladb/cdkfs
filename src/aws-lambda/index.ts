export * from './lambda-function';
export * from './auto-discover';