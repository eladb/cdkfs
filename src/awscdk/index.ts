export * from './auto-discover';
export * from './awscdk-app-ts';
export * from './awscdk-construct';
export * from './awscdk-deps';
export * from './cdk-config';
export * from './cdk-tasks';
export * from './integration-test';
export * from './java-app';
export * from './lambda-function';
