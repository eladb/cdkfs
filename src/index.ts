export * from './file';
export * from './github-workflow';
export * from './ignore-file';
export * from './jsii-project';
export * from './json';
export * from './license';
export * from './node-project';
export * from './project';
export * from './component';
export * from './semver';
export * from './version';
export * from './eslint';
export * from './jest';
export * from './typescript';
export * from './mergify';
export * from './dependabot';