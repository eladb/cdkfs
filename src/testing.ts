import { synthSnapshot } from './_test-utils';
import { Project } from "./project";

export class Testing {
  public static synth(project: Project): Record<string, any> {
    return synthSnapshot(project);
  }

  private constructor() {} // utility
}