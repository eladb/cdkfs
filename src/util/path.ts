/**
 * Helper function to format a path as dot notation regardless of how it
 * was handed in.
 *
 * @param path - can be formatted as "path", "./path", or "/path"
 * @returns "./path"
 */
export function formatPathAsDotNotation(path: string): string {
  if (path.startsWith(".")) {
    return path;
  }

  if (path.startsWith("/")) {
    return `.${path}`;
  }

  return `./${path}`;
}
