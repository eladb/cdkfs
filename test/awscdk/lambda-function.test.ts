import { mkdirSync, writeFileSync } from 'fs';
import { join } from 'path';
import { TypeScriptProject } from '../../src';
import * as awscdk from '../../src/awscdk';
import { Testing } from '../../src/testing';

describe('bundled function', () => {
  let generatedSource: string;
  let tasks: Record<string, any>;

  beforeEach(() => {
    const project = new TypeScriptProject({ name: 'hello', defaultReleaseBranch: 'main' });

    new awscdk.LambdaFunction(project, {
      entrypoint: join('src', 'hello.lambda.ts'),
      srcdir: project.srcdir,
      libdir: project.libdir,
    });

    const snapshot = Testing.synth(project);
    generatedSource = snapshot['src/hello-function.ts'];
    tasks = snapshot['.projen/tasks.json'].tasks;
  });

  test('generates source code for a lambda construct', () => {
    expect(generatedSource).toMatchSnapshot();
  });

  test('creates a single project-wide bundle task', () => {
    expect(tasks.bundle).toEqual({
      description: 'Bundle assets',
      name: 'bundle',
      steps: [{ spawn: 'bundle:hello' }],
    });
  });

  test('creates a specific bundle task for this function', () => {
    expect(tasks['bundle:hello']).toEqual({
      description: 'Create a JavaScript bundle from src/hello.lambda.ts',
      name: 'bundle:hello',
      steps: [
        {
          exec: 'esbuild --bundle src/hello.lambda.ts --target="node14" --platform="node" --outfile="lib/hello.lambda.bundle/index.js" --external:aws-sdk --sourcemap',
        },
      ],
    });
  });

  test('spawns the bundle task as part of pre-compilation', () => {
    expect(tasks['pre-compile']).toEqual({
      description: 'Prepare the project for compilation',
      name: 'pre-compile',
      steps: [
        { spawn: 'bundle' },
      ],
    });
  });
});

test('fails if entrypoint does not have the .lambda suffix', () => {
  const project = new TypeScriptProject({ name: 'hello', defaultReleaseBranch: 'main' });
  expect(() => new awscdk.LambdaFunction(project, {
    entrypoint: join('src', 'hello-no-lambda.ts'),
    libdir: project.libdir,
    srcdir: project.srcdir,
  })).toThrow('hello-no-lambda.ts must have a .lambda.ts extension');
});

test('fails if entrypoint is not under the source tree', () => {
  const project = new TypeScriptProject({ name: 'hello', defaultReleaseBranch: 'main' });
  expect(() => new awscdk.LambdaFunction(project, {
    entrypoint: join('boom', 'hello-no-lambda.ts'),
    libdir: project.libdir,
    srcdir: project.srcdir,
  })).toThrow('boom/hello-no-lambda.ts must be under src');
});

test('constructFile and constructName can be used to customize the generated construct', () => {
  const project = new TypeScriptProject({ name: 'hello', defaultReleaseBranch: 'main' });

  new awscdk.LambdaFunction(project, {
    entrypoint: join('src', 'hello.lambda.ts'),
    constructFile: 'my-construct.ts',
    constructName: 'MyConstruct',
    libdir: project.libdir,
    srcdir: project.srcdir,
  });

  const snapshot = Testing.synth(project);
  const generatedSource = snapshot['src/my-construct.ts'];
  expect(generatedSource).toMatchSnapshot();
});

test('runtime can be used to customize the lambda runtime and esbuild target', () => {
  const project = new TypeScriptProject({ name: 'hello', defaultReleaseBranch: 'main' });

  new awscdk.LambdaFunction(project, {
    entrypoint: join('src', 'hello.lambda.ts'),
    runtime: awscdk.LambdaRuntime.NODEJS_12_X,
    libdir: project.libdir,
    srcdir: project.srcdir,
  });

  const snapshot = Testing.synth(project);
  const generatedSource = snapshot['src/hello-function.ts'];
  const tasks = snapshot['.projen/tasks.json'].tasks;
  expect(generatedSource).toContain('runtime: lambda.Runtime.NODEJS_12_X,');
  expect(tasks['bundle:hello']).toEqual({
    description: 'Create a JavaScript bundle from src/hello.lambda.ts',
    name: 'bundle:hello',
    steps: [
      {
        exec: 'esbuild --bundle src/hello.lambda.ts --target="node12" --platform="node" --outfile="lib/hello.lambda.bundle/index.js" --external:aws-sdk --sourcemap',
      },
    ],
  });
});

test('eslint allows handlers to import dev dependencies', () => {
  const project = new TypeScriptProject({ name: 'hello', defaultReleaseBranch: 'main' });
  new awscdk.LambdaFunction(project, { entrypoint: join('src', 'hello.lambda.ts'), libdir: project.libdir, srcdir: project.srcdir });
  new awscdk.LambdaFunction(project, { entrypoint: join('src', 'world.lambda.ts'), libdir: project.libdir, srcdir: project.srcdir });

  const snapshot = Testing.synth(project);
  expect(snapshot['.eslintrc.json'].rules['import/no-extraneous-dependencies']).toStrictEqual([
    'error', {
      devDependencies: ['**/test/**', '**/build-tools/**', 'src/hello.lambda.ts', 'src/world.lambda.ts'],
      optionalDependencies: false,
      peerDependencies: true,
    },
  ]);
});

test('esbuild dependency is added', () => {
  const project = new TypeScriptProject({ name: 'hello', defaultReleaseBranch: 'main' });
  new awscdk.LambdaFunction(project, { entrypoint: join('src', 'hello.lambda.ts'), libdir: project.libdir, srcdir: project.srcdir });
  new awscdk.LambdaFunction(project, { entrypoint: join('src', 'world.lambda.ts'), libdir: project.libdir, srcdir: project.srcdir });

  const snapshot = Testing.synth(project);
  const deps = snapshot['.projen/deps.json'].dependencies;
  expect(deps.filter((d: any) => d.name === 'esbuild').length).toEqual(1);
});

test('multiple functions', () => {
  const project = new TypeScriptProject({ name: 'hello', defaultReleaseBranch: 'main' });
  new awscdk.LambdaFunction(project, { entrypoint: join('src', 'hello.lambda.ts'), libdir: project.libdir, srcdir: project.srcdir });
  new awscdk.LambdaFunction(project, { entrypoint: join('src', 'world.lambda.ts'), libdir: project.libdir, srcdir: project.srcdir });

  const snapshot = Testing.synth(project);
  expect(snapshot['src/hello-function.ts']).toMatchSnapshot();
  expect(snapshot['src/world-function.ts']).toMatchSnapshot();
});

test('auto-discover', () => {
  const project = new TypeScriptProject({ name: 'hello', defaultReleaseBranch: 'main' });

  // create fake handlers
  const srcdir = join(project.outdir, project.srcdir);
  mkdirSync(join(srcdir, 'subdir'), { recursive: true });
  writeFileSync(join(srcdir, 'hello.lambda.ts'), 'export function handler() {}');
  writeFileSync(join(srcdir, 'subdir', 'world.lambda.ts'), 'export function handler() {}');
  writeFileSync(join(srcdir, 'subdir', 'jangy.lambda.ts'), 'export function handler() {}');

  new awscdk.AutoDiscover(project, {
    libdir: project.libdir,
    srcdir: project.srcdir,
    lambdaOptions: {
      runtime: awscdk.LambdaRuntime.NODEJS_12_X,
    },
  });

  const snapshot = Testing.synth(project);
  expect(snapshot['src/hello-function.ts']).toMatchSnapshot();
  expect(snapshot['src/subdir/world-function.ts']).toMatchSnapshot();
  expect(snapshot['src/subdir/jangy-function.ts']).toMatchSnapshot();
  const tasks = snapshot['.projen/tasks.json'].tasks;
  expect(tasks.bundle).toMatchSnapshot();

  const expected = [
    'bundle:hello',
    'bundle:hello:watch',
    'bundle:subdir/jangy',
    'bundle:subdir/jangy:watch',
    'bundle:subdir/world',
    'bundle:subdir/world:watch',
  ];

  for (const name of expected) {
    expect(tasks[name]).not.toBeUndefined();
    expect(tasks[name]).toMatchSnapshot();
  }
});