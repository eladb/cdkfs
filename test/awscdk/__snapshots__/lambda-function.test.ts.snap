// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auto-discover 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".
import * as path from 'path';
import * as lambda from '@aws-cdk/aws-lambda';
import { Construct } from '@aws-cdk/core';

/**
 * Props for HelloFunction
 */
export interface HelloFunctionProps extends lambda.FunctionOptions {
}

/**
 * An AWS Lambda function which executes hello.
 */
export class HelloFunction extends lambda.Function {
  constructor(scope: Construct, id: string, props?: HelloFunctionProps) {
    super(scope, id, {
      description: 'hello.lambda.ts',
      ...props,
      runtime: lambda.Runtime.NODEJS_12_X,
      handler: 'index.handler',
      code: lambda.Code.fromAsset(path.join(__dirname, 'hello.bundle')),
    });
  }
}"
`;

exports[`auto-discover 2`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".
import * as path from 'path';
import * as lambda from '@aws-cdk/aws-lambda';
import { Construct } from '@aws-cdk/core';

/**
 * Props for WorldFunction
 */
export interface WorldFunctionProps extends lambda.FunctionOptions {
}

/**
 * An AWS Lambda function which executes subdir/world.
 */
export class WorldFunction extends lambda.Function {
  constructor(scope: Construct, id: string, props?: WorldFunctionProps) {
    super(scope, id, {
      description: 'subdir/world.lambda.ts',
      ...props,
      runtime: lambda.Runtime.NODEJS_12_X,
      handler: 'index.handler',
      code: lambda.Code.fromAsset(path.join(__dirname, 'subdir/world.bundle')),
    });
  }
}"
`;

exports[`auto-discover 3`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".
import * as path from 'path';
import * as lambda from '@aws-cdk/aws-lambda';
import { Construct } from '@aws-cdk/core';

/**
 * Props for JangyFunction
 */
export interface JangyFunctionProps extends lambda.FunctionOptions {
}

/**
 * An AWS Lambda function which executes subdir/jangy.
 */
export class JangyFunction extends lambda.Function {
  constructor(scope: Construct, id: string, props?: JangyFunctionProps) {
    super(scope, id, {
      description: 'subdir/jangy.lambda.ts',
      ...props,
      runtime: lambda.Runtime.NODEJS_12_X,
      handler: 'index.handler',
      code: lambda.Code.fromAsset(path.join(__dirname, 'subdir/jangy.bundle')),
    });
  }
}"
`;

exports[`auto-discover 4`] = `
Object {
  "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
  "env": Object {
    "PATH": "$(npx -c \\"node -e \\\\\\"console.log(process.env.PATH)\\\\\\"\\")",
  },
  "tasks": Object {
    "build": Object {
      "description": "Full release build (test+compile)",
      "name": "build",
      "steps": Array [
        Object {
          "exec": "npx projen",
        },
        Object {
          "spawn": "test",
        },
        Object {
          "spawn": "compile",
        },
        Object {
          "spawn": "package",
        },
      ],
    },
    "bump": Object {
      "condition": "! git log --oneline -1 | grep -q \\"chore(release):\\"",
      "description": "Bumps version based on latest git tag and generates a changelog entry",
      "env": Object {
        "BUMPFILE": "dist/version.txt",
        "CHANGELOG": "dist/changelog.md",
        "OUTFILE": "package.json",
        "RELEASETAG": "dist/releasetag.txt",
      },
      "name": "bump",
      "steps": Array [
        Object {
          "builtin": "release/bump-version",
        },
      ],
    },
    "bundle": Object {
      "description": "Bundle assets",
      "name": "bundle",
      "steps": Array [
        Object {
          "spawn": "bundle:hello",
        },
        Object {
          "spawn": "bundle:subdir/jangy",
        },
        Object {
          "spawn": "bundle:subdir/world",
        },
      ],
    },
    "bundle:hello": Object {
      "description": "Create a JavaScript bundle from src/hello.lambda.ts",
      "name": "bundle:hello",
      "steps": Array [
        Object {
          "exec": "esbuild --bundle src/hello.lambda.ts --target=\\"node12\\" --platform=\\"node\\" --outfile=\\"lib/hello.bundle/index.js\\" --external:aws-sdk",
        },
      ],
    },
    "bundle:subdir/jangy": Object {
      "description": "Create a JavaScript bundle from src/subdir/jangy.lambda.ts",
      "name": "bundle:subdir/jangy",
      "steps": Array [
        Object {
          "exec": "esbuild --bundle src/subdir/jangy.lambda.ts --target=\\"node12\\" --platform=\\"node\\" --outfile=\\"lib/subdir/jangy.bundle/index.js\\" --external:aws-sdk",
        },
      ],
    },
    "bundle:subdir/world": Object {
      "description": "Create a JavaScript bundle from src/subdir/world.lambda.ts",
      "name": "bundle:subdir/world",
      "steps": Array [
        Object {
          "exec": "esbuild --bundle src/subdir/world.lambda.ts --target=\\"node12\\" --platform=\\"node\\" --outfile=\\"lib/subdir/world.bundle/index.js\\" --external:aws-sdk",
        },
      ],
    },
    "clobber": Object {
      "condition": "git diff --exit-code > /dev/null",
      "description": "hard resets to HEAD of origin and cleans the local repo",
      "env": Object {
        "BRANCH": "$(git branch --show-current)",
      },
      "name": "clobber",
      "steps": Array [
        Object {
          "exec": "git checkout -b scratch",
          "name": "save current HEAD in \\"scratch\\" branch",
        },
        Object {
          "exec": "git checkout $BRANCH",
        },
        Object {
          "exec": "git fetch origin",
          "name": "fetch latest changes from origin",
        },
        Object {
          "exec": "git reset --hard origin/$BRANCH",
          "name": "hard reset to origin commit",
        },
        Object {
          "exec": "git clean -fdx",
          "name": "clean all untracked files",
        },
        Object {
          "say": "ready to rock! (unpushed commits are under the \\"scratch\\" branch)",
        },
      ],
    },
    "compile": Object {
      "description": "Only compile",
      "name": "compile",
      "steps": Array [
        Object {
          "exec": "tsc --build",
        },
        Object {
          "spawn": "bundle",
        },
      ],
    },
    "default": Object {
      "name": "default",
      "steps": Array [
        Object {
          "exec": "node .projenrc.js",
        },
      ],
    },
    "eslint": Object {
      "description": "Runs eslint against the codebase",
      "name": "eslint",
      "steps": Array [
        Object {
          "exec": "eslint --ext .ts,.tsx --fix --no-error-on-unmatched-pattern src test build-tools .projenrc.js",
        },
      ],
    },
    "package": Object {
      "description": "Create an npm tarball",
      "name": "package",
      "steps": Array [
        Object {
          "exec": "mkdir -p dist/js",
        },
        Object {
          "exec": "yarn pack",
        },
        Object {
          "exec": "mv *.tgz dist/js/",
        },
      ],
    },
    "publish:github": Object {
      "description": "Publish this package to GitHub Releases",
      "name": "publish:github",
      "requiredEnv": Array [
        "GITHUB_TOKEN",
        "GITHUB_REPOSITORY",
        "GITHUB_REF",
      ],
      "steps": Array [
        Object {
          "exec": "errout=$(mktemp); gh release create $(cat dist/releasetag.txt) -R $GITHUB_REPOSITORY -F dist/changelog.md -t $(cat dist/releasetag.txt) --target $GITHUB_REF 2> $errout && true; exitcode=$?; if [ $exitcode -ne 0 ] && ! grep -q \\"Release.tag_name already exists\\" $errout; then cat $errout; exit $exitcode; fi",
        },
      ],
    },
    "release": Object {
      "description": "Prepare a release from \\"main\\" branch",
      "env": Object {
        "RELEASE": "true",
      },
      "name": "release",
      "steps": Array [
        Object {
          "exec": "rm -fr dist",
        },
        Object {
          "spawn": "bump",
        },
        Object {
          "spawn": "build",
        },
        Object {
          "spawn": "unbump",
        },
        Object {
          "exec": "git diff --ignore-space-at-eol --exit-code",
        },
      ],
    },
    "test": Object {
      "description": "Run tests",
      "name": "test",
      "steps": Array [
        Object {
          "exec": "rm -fr lib/",
        },
        Object {
          "exec": "jest --passWithNoTests --all --updateSnapshot",
        },
        Object {
          "spawn": "eslint",
        },
      ],
    },
    "test:update": Object {
      "description": "Update jest snapshots",
      "name": "test:update",
      "steps": Array [
        Object {
          "exec": "jest --updateSnapshot",
        },
      ],
    },
    "test:watch": Object {
      "description": "Run jest in watch mode",
      "name": "test:watch",
      "steps": Array [
        Object {
          "exec": "jest --watch",
        },
      ],
    },
    "unbump": Object {
      "description": "Restores version to 0.0.0",
      "env": Object {
        "BUMPFILE": "dist/version.txt",
        "CHANGELOG": "dist/changelog.md",
        "OUTFILE": "package.json",
        "RELEASETAG": "dist/releasetag.txt",
      },
      "name": "unbump",
      "steps": Array [
        Object {
          "builtin": "release/reset-version",
        },
      ],
    },
    "upgrade": Object {
      "description": "upgrade dependencies",
      "env": Object {
        "CI": "0",
      },
      "name": "upgrade",
      "steps": Array [
        Object {
          "exec": "npm-check-updates --dep dev --upgrade --target=minor --reject='projen'",
        },
        Object {
          "exec": "npm-check-updates --dep optional --upgrade --target=minor --reject='projen'",
        },
        Object {
          "exec": "npm-check-updates --dep peer --upgrade --target=minor --reject='projen'",
        },
        Object {
          "exec": "npm-check-updates --dep prod --upgrade --target=minor --reject='projen'",
        },
        Object {
          "exec": "npm-check-updates --dep bundle --upgrade --target=minor --reject='projen'",
        },
        Object {
          "exec": "yarn install --check-files",
        },
        Object {
          "exec": "yarn upgrade @types/jest @types/node @typescript-eslint/eslint-plugin @typescript-eslint/parser esbuild eslint eslint-import-resolver-node eslint-import-resolver-typescript eslint-plugin-import jest jest-junit json-schema npm-check-updates standard-version ts-jest typescript",
        },
        Object {
          "exec": "npx projen",
        },
      ],
    },
    "upgrade-projen": Object {
      "description": "upgrade projen",
      "env": Object {
        "CI": "0",
      },
      "name": "upgrade-projen",
      "steps": Array [
        Object {
          "exec": "npm-check-updates --dep dev --upgrade --target=minor --filter='projen'",
        },
        Object {
          "exec": "npm-check-updates --dep optional --upgrade --target=minor --filter='projen'",
        },
        Object {
          "exec": "npm-check-updates --dep peer --upgrade --target=minor --filter='projen'",
        },
        Object {
          "exec": "npm-check-updates --dep prod --upgrade --target=minor --filter='projen'",
        },
        Object {
          "exec": "npm-check-updates --dep bundle --upgrade --target=minor --filter='projen'",
        },
        Object {
          "exec": "yarn install --check-files",
        },
        Object {
          "exec": "yarn upgrade projen",
        },
        Object {
          "exec": "npx projen",
        },
      ],
    },
    "watch": Object {
      "description": "Watch & compile in the background",
      "name": "watch",
      "steps": Array [
        Object {
          "exec": "tsc --build -w",
        },
      ],
    },
  },
}
`;

exports[`bundled function generates source code for a lambda construct 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".
import * as path from 'path';
import * as lambda from '@aws-cdk/aws-lambda';
import { Construct } from '@aws-cdk/core';

/**
 * Props for HelloFunction
 */
export interface HelloFunctionProps extends lambda.FunctionOptions {
}

/**
 * An AWS Lambda function which executes hello.
 */
export class HelloFunction extends lambda.Function {
  constructor(scope: Construct, id: string, props?: HelloFunctionProps) {
    super(scope, id, {
      description: 'hello.lambda.ts',
      ...props,
      runtime: lambda.Runtime.NODEJS_14_X,
      handler: 'index.handler',
      code: lambda.Code.fromAsset(path.join(__dirname, 'hello.bundle')),
    });
  }
}"
`;

exports[`constructFile and constructName can be used to customize the generated construct 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".
import * as path from 'path';
import * as lambda from '@aws-cdk/aws-lambda';
import { Construct } from '@aws-cdk/core';

/**
 * Props for MyConstruct
 */
export interface MyConstructProps extends lambda.FunctionOptions {
}

/**
 * An AWS Lambda function which executes hello.
 */
export class MyConstruct extends lambda.Function {
  constructor(scope: Construct, id: string, props?: MyConstructProps) {
    super(scope, id, {
      description: 'hello.lambda.ts',
      ...props,
      runtime: lambda.Runtime.NODEJS_14_X,
      handler: 'index.handler',
      code: lambda.Code.fromAsset(path.join(__dirname, 'hello.bundle')),
    });
  }
}"
`;

exports[`multiple functions 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".
import * as path from 'path';
import * as lambda from '@aws-cdk/aws-lambda';
import { Construct } from '@aws-cdk/core';

/**
 * Props for HelloFunction
 */
export interface HelloFunctionProps extends lambda.FunctionOptions {
}

/**
 * An AWS Lambda function which executes hello.
 */
export class HelloFunction extends lambda.Function {
  constructor(scope: Construct, id: string, props?: HelloFunctionProps) {
    super(scope, id, {
      description: 'hello.lambda.ts',
      ...props,
      runtime: lambda.Runtime.NODEJS_14_X,
      handler: 'index.handler',
      code: lambda.Code.fromAsset(path.join(__dirname, 'hello.bundle')),
    });
  }
}"
`;

exports[`multiple functions 2`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".
import * as path from 'path';
import * as lambda from '@aws-cdk/aws-lambda';
import { Construct } from '@aws-cdk/core';

/**
 * Props for WorldFunction
 */
export interface WorldFunctionProps extends lambda.FunctionOptions {
}

/**
 * An AWS Lambda function which executes world.
 */
export class WorldFunction extends lambda.Function {
  constructor(scope: Construct, id: string, props?: WorldFunctionProps) {
    super(scope, id, {
      description: 'world.lambda.ts',
      ...props,
      runtime: lambda.Runtime.NODEJS_14_X,
      handler: 'index.handler',
      code: lambda.Code.fromAsset(path.join(__dirname, 'world.bundle')),
    });
  }
}"
`;
